pkg/river/json_walker.go

```diff
@@ -0,0 +1,409 @@
+package river
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/grafana/agent/pkg/river/token/builder"
+
+	"github.com/grafana/agent/pkg/river/internal/rivertags"
+	"github.com/grafana/agent/pkg/river/internal/value"
+)
+
+// ComponentField represents a component in river.
+type ComponentField struct {
+	Field        `json:",omitempty"`
+	References   []string    `json:"referencesTo"`
+	ReferencedBy []string    `json:"referencedBy"`
+	Health       *Health     `json:"health"`
+	Original     string      `json:"original"`
+	Arguments    interface{} `json:"arguments,omitempty"`
+	Exports      interface{} `json:"exports,omitempty"`
+	DebugInfo    interface{} `json:"debugInfo,omitempty"`
+}
+
+// Field represents a value in river.
+type Field struct {
+	ID    string      `json:"id,omitempty"`
+	Key   string      `json:"key,omitempty"`
+	Label string      `json:"label,omitempty"`
+	Name  string      `json:"name,omitempty"`
+	Type  string      `json:"type,omitempty"`
+	Value interface{} `json:"value,omitempty"`
+	Body  []*Field    `json:"body,omitempty"`
+}
+
+// Health represents the health of a component.
+type Health struct {
+	State       string    `json:"state"`
+	Message     string    `json:"message"`
+	UpdatedTime time.Time `json:"updatedTime"`
+}
```

Comment:
I would prefer avoiding adding Flow-specific pieces into the River code and we can decouple the two here. The code here can force on creating a JSON value for a River block, body, or value, and then the rest can live in pkg/flow.

Reply:
To add a little extra flavor on _why_ I think we should decouple things: IMO this introduces a "soft" cyclic dependency, where `pkg/flow` imports `pkg/river` but `pkg/river` has Flow-specific concepts like component components.

IMO it makes the separation of concerns a little blurry and might make it harder to reason about code as we continue adding new functionality.

Reply:
Moved things around, so the actual river json only knows about the most basic building blocks: blocks and attrs.